### ISSUE TITLE
Lint to detect non `#[repr(C)]` type in `#[repr(C)]` struct

### ISSUE THREAD

**Issue by xFrednet** [rust-lang/rust-clippy#9357]
### What it does

Rust doesn't have a stable ABI, to still allow the exchange of data with external code an ABI can be defined per type. This is done with the `#[repr(abi)]` attribute. However, (AFAIK), the attribute only affects the type itself, meaning that the user has to ensure that all fields in the type also have `#[repr(abi)]`.

The lint should warn, if a struct has an `#[repr(abi)]` attribute, but uses a type which doesn't have `#[repr(abi)]`. Generics should probably be ignored, as we can't check for those ABIs.

### Lint Name

field_without_repr

### Category

restriction

### Advantage

Safe communication with other code

### Drawbacks

1. For external types, it's hard to impossible to create a suggestion.
2. There are reasons to have `#[repr(C)]` on a type, but not on each field. This will therefore create FPs
3. Maybe a lot of lint triggers

These drawbacks are fine if the lint is in the lint is in the `restriction` category

### Example

```rs
// This is fine
#[repr(C)]
struct Magic {
    data: u32
}

// This can cause problems
#[repr(C)]
struct Wizzard {
    // `String` is not `#[repr(C)]`
    name: String,
}
```

**Comment by leonardo-m** [comment]
Sounds important.
